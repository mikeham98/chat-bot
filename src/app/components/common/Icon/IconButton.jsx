import React from 'react';
import styles from '../../../../themes/components/icon.scss';
import Icon from "./index";

const IconButton = ({onClick, type, icon, style, color, hoverColor, height, width, disabled}) => {
    return (
        <button
            type={type}
            onClick={onClick}
            disabled={disabled}
            style={style}
            className={styles.iconButton}
        >
            <Icon
                IconComponent={icon}
                color={color}
                hoverColor={hoverColor}
                height={height}
                width={width}
            />
        </button>
    );
};

export default React.memo(IconButton);