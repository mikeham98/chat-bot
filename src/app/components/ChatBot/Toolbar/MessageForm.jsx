import React from 'react';
import PropTypes from "prop-types";
import {Formik} from 'formik';
import MessageInput from './MessageInput';
import IconButton from "../../common/Icon/IconButton";
import sendIcon from '../../../../../assets/icons/rightArrow.svg';
import styles from '../../../../themes/components/toolbar.scss';

class MessageForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit({message}, {setSubmitting, resetForm}) {
        const callback = () => {
            setSubmitting(false);
            resetForm()
        };
        this.props.sendMessage(message, callback);
    }

    validate(values) {
        let errors = {};
        if (!values.message) {
            errors.message = 'Required';
        }
        return errors;
    }

    render() {
        return (
            <Formik
                initialValues={this.props.initialValues}
                onSubmit={this.handleSubmit}
                validate={this.validate}
            >
                {formikProps => {
                    const {
                        values,
                        isSubmitting,
                        handleChange,
                        handleBlur,
                        handleSubmit,
                    } = formikProps;
                    return (
                        <form className={styles.messageFormWrapper} onSubmit={handleSubmit}>
                            <MessageInput
                                value={values.message}
                                onChange={handleChange}
                                onBlur={handleBlur}
                            />
                            <IconButton
                                type='submit'
                                icon={sendIcon}
                                style={{margin: '0 10px'}}
                                color={(values.message && values.message.length) ? '#6872FF' : '#939aab'}
                                height='30px'
                                width='30px'
                                disabled={isSubmitting}
                            />
                        </form>
                    );
                }}
            </Formik>
        )
    }
}

MessageForm.defaultProps = {
    initialValues: {
        message: ''
    }
};

MessageForm.propTypes = {
    initialValues: PropTypes.shape({
        message: PropTypes.string
    }),
    sendMessage: PropTypes.func.isRequired
};

export default MessageForm;