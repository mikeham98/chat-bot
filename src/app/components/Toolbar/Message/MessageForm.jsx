import React from 'react';
import {Formik} from 'formik';
import MessageInput from './MessageInput';
import IconButton from "../../common/Icon/Button";
import sendLogo from '../../../../../assets/icons/right-arrow-1.svg';
import styles from '../../../../themes/toolbar.scss';

export default class MessageForm extends React.Component {
    constructor(props) {
        super(props);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit({message}, {setSubmitting, resetForm}) {
        const callback = () => {
            setSubmitting(false);
            resetForm()
        };
        this.props.sendMessage(message, callback);
    }

    render() {
        return (
            <Formik
                initialValues={this.props.initialValues}
                onSubmit={this.handleSubmit}
                // validationSchema={Yup.object().shape({
                //     email: Yup.string()
                //         .email()
                //         .required('Required'),
                // })}
            >
                {formikProps => {
                    const {
                        values,
                        touched,
                        errors,
                        isSubmitting,
                        handleChange,
                        handleBlur,
                        handleSubmit,
                    } = formikProps;
                    return (
                        <form className={styles.messageFormWrapper} onSubmit={handleSubmit}>
                            <MessageInput
                                value={values.message}
                                onChange={handleChange}
                                onBlur={handleBlur}
                                touched={touched.message}
                                error={errors.message}
                            />
                            <IconButton
                                type='submit'
                                icon={sendLogo}
                                style={{margin: '0 10px'}}
                                color={values.message.length ? '#172B4D' : '#939aab'}
                                height='30px'
                                width='30px'
                                disabled={isSubmitting}
                            />
                        </form>
                    );
                }}
            </Formik>
        )
    }
}

MessageForm.defaultProps = {
    initialValues: {
        message: ''
    }
};