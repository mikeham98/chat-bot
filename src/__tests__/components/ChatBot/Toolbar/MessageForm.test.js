import React from 'react';
import {mount} from 'enzyme';
import MessageForm from '../../../../app/components/ChatBot/Toolbar/MessageForm';
import IconButton from "../../../../app/components/common/Icon/IconButton";

jest.mock("../../../../app/components/common/Icon/index.jsx", () => {
    return () => <div>Icon</div>
});

describe('MessageForm', () => {
    let mountedWrapper;
    let props = {};
    const wrapper = () => {
        if (!mountedWrapper) {
            mountedWrapper = mount(<MessageForm {...props}/>);
        }
        return mountedWrapper;
    };

    beforeEach(() => {
        props = {
            initialValues: {message: ''},
            sendMessage: jest.fn((message, callback) => callback())
        };
        mountedWrapper = undefined;
    });

    describe('render', () => {
        it('should return a form component with a MessageInput and an IconButton', () => {
            expect(wrapper()).toMatchSnapshot();
        });
        describe('IconButton', () => {
            it('should have color prop set to #939aab', () => {
                expect(wrapper().find(IconButton).first().props().color).toBe('#939aab');
            });
            it('should have color prop set to #172B4D because messages has a length', () => {
                wrapper().find('input').first().simulate('change', {
                    target: {
                        name: 'message',
                        value: 'this is a message'
                    }
                });
                wrapper().update();
                expect(wrapper().find(IconButton).first().props().color).toBe('#172B4D');
            });
        });
    });
    describe('functionality', () => {
        describe('validate', () => {
            it('should return an object with message set to Required', () => {
                const values = {
                    message: ''
                };
                const error = {
                    message: 'Required'
                };
                expect(wrapper().find('Formik').first().props().validate(values)).toEqual(error);
            });
            it('should return an empty object', () => {
                const values = {
                    message: 'this is a message'
                };
                const error = {};
                expect(wrapper().find('Formik').first().props().validate(values)).toEqual(error);
            });
        });
        describe('props', () => {
            describe('sendMessage', () => {
                it('should be called onSubmit along with setSubmitting and resetForm', () => {
                    const values = {
                        message: 'this is a message'
                    };
                    const formikBag = {
                        setSubmitting: jest.fn(),
                        resetForm: jest.fn(),
                    };
                    expect(props.sendMessage).not.toHaveBeenCalled();
                    expect(formikBag.setSubmitting).not.toHaveBeenCalled();
                    expect(formikBag.resetForm).not.toHaveBeenCalled();
                    wrapper().find('Formik').first().props().onSubmit(values, formikBag);
                    expect(props.sendMessage).toHaveBeenCalledTimes(1);
                    expect(props.sendMessage).toHaveBeenCalledWith(values.message, expect.any(Function));
                    expect(formikBag.resetForm).toHaveBeenCalledTimes(1);
                    expect(formikBag.setSubmitting).toHaveBeenCalledTimes(1);
                });
            });
        });
    });
});