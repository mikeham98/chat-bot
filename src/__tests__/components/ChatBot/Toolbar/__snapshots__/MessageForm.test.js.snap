// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm render should return a form component with a MessageInput and an IconButton 1`] = `
<MessageForm
  initialValues={
    Object {
      "message": "",
    }
  }
  sendMessage={[MockFunction]}
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "message": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <form
      className="messageFormWrapper"
      onSubmit={[Function]}
    >
      <Memo(MessageInput)
        onBlur={[Function]}
        onChange={[Function]}
        value=""
      >
        <input
          className="messageInputWrapper"
          name="message"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Type a message..."
          type="text"
          value=""
        />
      </Memo(MessageInput)>
      <Memo(IconButton)
        color="#939aab"
        disabled={false}
        height="30px"
        icon={
          Object {
            "canInstrument": true,
            "createTransformer": [Function],
            "getCacheKey": [Function],
            "process": [Function],
          }
        }
        style={
          Object {
            "margin": "0 10px",
          }
        }
        type="submit"
        width="30px"
      >
        <button
          className="iconButton"
          disabled={false}
          style={
            Object {
              "margin": "0 10px",
            }
          }
          type="submit"
        >
          <Component
            IconComponent={
              Object {
                "canInstrument": true,
                "createTransformer": [Function],
                "getCacheKey": [Function],
                "process": [Function],
              }
            }
            color="#939aab"
            height="30px"
            width="30px"
          >
            <div>
              Icon
            </div>
          </Component>
        </button>
      </Memo(IconButton)>
    </form>
  </Formik>
</MessageForm>
`;
